cmd/config.go

```diff
@@ -208,56 +208,37 @@ func mustGetHostConfig(alias string) *aliasConfigV10 {
 	return aliasCfg
 }
 
-var (
-	hostKeys      = regexp.MustCompile("^(https?://)(.*?):(.*?)@(.*?)$")
-	hostKeyTokens = regexp.MustCompile("^(https?://)(.*?):(.*?):(.*?)@(.*?)$")
-)
-
-// parse url usually obtained from env.
-func parseEnvURLStr(envURL string) (*url.URL, string, string, string, *probe.Error) {
-	var accessKey, secretKey, sessionToken string
-	var parsedURL string
-	if hostKeyTokens.MatchString(envURL) {
-		parts := hostKeyTokens.FindStringSubmatch(envURL)
-		if len(parts) != 6 {
-			return nil, "", "", "", errInvalidArgument().Trace(envURL)
-		}
-		accessKey = parts[2]
-		secretKey = parts[3]
-		sessionToken = parts[4]
-		parsedURL = fmt.Sprintf("%s%s", parts[1], parts[5])
-	} else if hostKeys.MatchString(envURL) {
-		parts := hostKeys.FindStringSubmatch(envURL)
-		if len(parts) != 5 {
-			return nil, "", "", "", errInvalidArgument().Trace(envURL)
-		}
-		accessKey = parts[2]
-		secretKey = parts[3]
-		parsedURL = fmt.Sprintf("%s%s", parts[1], parts[4])
-	}
-	var u *url.URL
-	var e error
-	if parsedURL != "" {
-		u, e = url.Parse(parsedURL)
-	} else {
-		u, e = url.Parse(envURL)
-	}
+// Parse MC_HOST url obtained from the environment
+// - It supports the following formats:
+//      http(s)://access-key:secret-key@host:port/
+//    AND
+///     http(s)://access-key:secret-key:session-token@host:port/
+// - Users need to manually escape any odds characters (except for @)
+//   in order for this to work correctly.
+func parseEnvURLStr(envURL string) (endpoint, accessKey, secretKey, sessionToken string, err *probe.Error) {
+	u, e := url.Parse(envURL)
 	if e != nil {
-		return nil, "", "", "", probe.NewError(e)
+		return "", "", "", "", probe.NewError(e)
 	}
+
+	endpoint = u.Scheme + "://" + u.Host
+	accessKey = u.User.Username()
+
+	pwd, _ := u.User.Password()
+	pwdFields := strings.Split(pwd, ":")
```

Comment:
You cannot assume that `password` doesn't have `:` @vadmeste - that's why the previous regex was aptly positional.

Reply:
> You cannot assume that password doesn't have : @vadmeste - that's why the previous regex was aptly positional.

Not sure I understand, if we cannot assume that the password doesn't have ':', how we will extract the session token from the url in this case ?